### ISSUE TITLE
Re-introducing the Space Layer

### ISSUE THREAD

**Issue by LynxAbraxas** [civctp2/civctp2#356]
This issue is meant for collecting information and ideas on how the space layer from CTP1 could be re-introduced in CTP2.

**Comment by LynxAbraxas** [comment]
As pointed out in the old documentation, most code from the CTP1 space layer was removed before the code was released:

https://github.com/civctp2/civctp2/blob/6b13c85a984909cf09aa3ac8bcb249d733a8e98b/doc/dev/include/a_faq_project.tex#L57-L60

Two cpp defines (`CTP1_HAS_RISEN_FROM_THE_GRAVE` and `_SMALL_MAPPOINTS`) concerning the space layer are:

https://github.com/civctp2/civctp2/blob/6b13c85a984909cf09aa3ac8bcb249d733a8e98b/doc/dev/include/a_defines.tex#L58
https://github.com/civctp2/civctp2/blob/6b13c85a984909cf09aa3ac8bcb249d733a8e98b/doc/dev/include/a_defines.tex#L64

Traces in the CTP2 code base I could find so far:

generally disabled:

https://github.com/civctp2/civctp2/blob/6b13c85a984909cf09aa3ac8bcb249d733a8e98b/ctp2_code/gs/gameobj/Player.cpp#L8686-L8694

https://github.com/civctp2/civctp2/blob/6b13c85a984909cf09aa3ac8bcb249d733a8e98b/ctp2_code/gs/gameobj/Unit.h#L462-L468

`CTP1_HAS_RISEN_FROM_THE_GRAVE`:

https://github.com/civctp2/civctp2/blob/6b13c85a984909cf09aa3ac8bcb249d733a8e98b/ctp2_code/ui/aui_ctp2/keypress.cpp#L995-L1006

`_SMALL_MAPPOINTS` at many places e.g. here:

https://github.com/civctp2/civctp2/blob/6b13c85a984909cf09aa3ac8bcb249d733a8e98b/ctp2_code/gs/world/MapPoint.h#L46

https://github.com/civctp2/civctp2/blob/6b13c85a984909cf09aa3ac8bcb249d733a8e98b/ctp2_code/gs/world/MapPoint.h#L80-L100

**Comment by LynxAbraxas** [comment]
In CTP1 `ctp/ctp_data/default/gamedata/terrain.txt` contains (after `TERRAIN_WATER_VOLCANO` and before `TERRAIN_WATER_SHELF`):

```
### TERRAIN DONE ###	0

TERRAIN_SPACE	{
TERRAIN_TYPE_SPACE	

TERRAIN_DEFAULT_ICON	ICON_TERRAIN_SPACE
```

which is missing in the corresponding CTP2 file:

https://github.com/civctp2/civctp2/blob/6b13c85a984909cf09aa3ac8bcb249d733a8e98b/ctp2_data/default/gamedata/terrain.txt#L786-L788


So far I was not able to find space "tiles" or some kind of "overlay tile" in the CTP1 `ctp/ctp_data/default/graphics/tiles/gtset565.til` with TileEdit. However, the readme of TileEdit mentions a "block of unknown meaning", which I could imagine could contain such "overlay tiles" for drawing the space layer similarly to e.g. transparent force-fields (#218).
The tile improvements for space can be found in `gtset565.til` of CTP1 and CTP2.

**Comment by LynxAbraxas** [comment]
@williamd67 as you are very well into the graphics of the game: 
Do you see a simple and especially fast way to re-implement the visual effect for the space layer? 
The simplest might be just to "whiten" the tiles of the map including all units not in space (i.e. with `MapPointData.z == 0`). 
In the linux port of CTP1, this seems to be done with interleaving (like it used to be done for force fields before #218) of some form of varying cloud tiles:
![ss_2021-07-30_23:52:37](https://user-images.githubusercontent.com/21012234/127715894-26b278aa-c197-40f8-b798-98533f2c2d1c.png)
![ss_2021-07-30_23:52:07](https://user-images.githubusercontent.com/21012234/127715893-bd3b042c-f61f-46b3-af26-be71245cb864.png)
For a start, we could go without cloud tiles but just with a plain whitening possibly with transparency instead of interleaving as a basic implementation for
https://github.com/civctp2/civctp2/blob/6b13c85a984909cf09aa3ac8bcb249d733a8e98b/ctp2_code/ui/aui_ctp2/keypress.cpp#L1001
 What do you think?

**Comment by LynxAbraxas** [comment]
Enabling the z-coord of `MapPoint`, i.e. not defining `_SMALL_MAPPOINTS` with a default of 0 for z compiles all fine and all tests except loading a game work as well:
https://gitlab.com/LynxAbraxas/civctp2/-/pipelines/352472823
The loading of games would need a converter version that loads 2D points and writes 3D points (with z= 0).

**Comment by LynxAbraxas** [comment]
> For a start, we could go without cloud tiles but just with a plain whitening possibly with transparency instead of interleaving as a basic implementation for
> 
> https://github.com/civctp2/civctp2/blob/6b13c85a984909cf09aa3ac8bcb249d733a8e98b/ctp2_code/ui/aui_ctp2/keypress.cpp#L1001


@williamd67 Where in the code would you recommend to introduce blending a transparent white into the tiles (similar to Your aafb5252)?
I.e. any better spot than:
https://github.com/civctp2/civctp2/blob/ae8b55597fd52a5f81a216377dc6bd9fce5e238f/ctp2_code/gfx/tilesys/tiledraw.cpp#L1709

**Comment by MartinGuehmann** [comment]
> So far I was not able to find space "tiles" or some kind of "overlay tile" in the CTP1 `ctp/ctp_data/default/graphics/tiles/gtset565.til` with TileEdit. However, the readme of TileEdit mentions a "block of unknown meaning", which I could imagine could contain such "overlay tiles" for drawing the space layer similarly to e.g. transparent force-fields (#218). The tile improvements for space can be found in `gtset565.til` of CTP1 and CTP2.

No, I think it refers to what we called the last block in the tile file as it is the last block. It is basically a set of tile rules of combining single tiles to something bigger, for instance mountain ranges or the volcano island in the Samurai scenario. The volcano also consists of more than one tile.

**Comment by LynxAbraxas** [comment]
> > So far I was not able to find space "tiles" or some kind of "overlay tile" in the CTP1 `ctp/ctp_data/default/graphics/tiles/gtset565.til` with TileEdit. However, the readme of TileEdit mentions a "block of unknown meaning", which I could imagine could contain such "overlay tiles" for drawing the space layer similarly to e.g. transparent force-fields (#218). The tile improvements for space can be found in `gtset565.til` of CTP1 and CTP2.
> 
> No, I think it refers to what we called the last block in the tile file as it is the last block. It is basically a set of tile rules of combining single tiles to something bigger, for instance mountain ranges or the volcano island in the Samurai scenario. The volcano also consists of more than one tile.

Yes, I included it here: https://github.com/civctp2/civctp2/pull/353/commits/3b25e43fc1d095830946e26d48e4db9ea9b3bd24

Independent of what was meant in the readme, do You have any idea where the "space overlay tiles" might be in CTP1's data files? Or do you recon achieving the cloud effect was hard coded?

**Comment by williamd67** [comment]
@LynxAbraxas, I expect that you want the `white` cloud on top of cities, units, goods, etc. In that case is DrawDitheredOverlayIntoMix not the right place.

I have experimented to introduce this in some other places and the best place I found is in `background_draw_handler` yet also there you will get side-effects as sometimes tiles will be rendered more than once meaning that the blending will happen more than once which will make the result opaque white.

Another element is that you probable want to draw space-elements `above` the white clouds. I have noticed that in the code there are drawing layers which as far as I have seen are not used, yet it may that they were introduced to draw the space layer. This approach will give you more control and prevent side-effects. I expect that you need to draw the whole map twice on different layers.

**Comment by MartinGuehmann** [comment]
> Independent of what was meant in the readme, do You have any idea where the "space overlay tiles" might be in CTP1's data files? Or do you recon achieving the cloud effect was hard coded?

From the screenshot, it is obvious that you have a complex cloud effect. However, I do not remember the details. Obviously, it is a complex pattern. Is it actually dynamic? Does it change over time? In that case, it could have been computed. Also what happens when you scroll does the pattern stay or does it move with the tiles?

The tile file does not contain anything, that looks like a graphical pattern of that, we should have seen it. This should be big, and Martin the Dane shared with me a little program that just interprets the bytes of the file as pixels, and so I could see something in it. And nothing of that was in.

**Comment by LynxAbraxas** [comment]
Many thanks @williamd67 for the infos and the hints on the unused layers that might come from drawing the space layer. I'll try to investigate that. If you don't mind, can you push your prototype implementation as a new branch to your fork of civctp2 such that I can take a look?
I wonder if it would be simplest to just draw one large semi-transparten white rectangle over the whole map if showing the space layer and then the space units on top of that.

**Comment by LynxAbraxas** [comment]
> From the screenshot, it is obvious that you have a complex cloud effect. However, I do not remember the details. Obviously, it is a complex pattern. Is it actually dynamic? Does it change over time? 

In the CTP1 Linux version the pattern is not dynamic or changes over time.

> Also what happens when you scroll does the pattern stay or does it move with the tiles?

It is static, i.e. stationary to the underlying land tiles.

I guess then we should just go for a semi-transparent white for the start and investigate a nicer visualization when all other things work.

**Comment by williamd67** [comment]
@LynxAbraxas for the space-layer experiment see [branch](https://github.com/williamd67/civctp2/tree/space-layer-prototype)

**Comment by LynxAbraxas** [comment]
Just for reference my branches that contain WIP on this:
https://github.com/LynxAbraxas/civctp2/tree/space-layer-prototype
https://github.com/LynxAbraxas/civctp2/tree/non-SMALL_MAPPOINTS
https://github.com/LynxAbraxas/civctp2/tree/non-SMALL_MAPPOINTS_converterVersion
